packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -406,7 +429,14 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
 
     // Temporarily exempt `newValue` from the maxLength limit if it has a
     // composing text going, until the composing is finished.
-    return newValue.composing.isValid ? newValue : truncate(newValue, maxLength);
+    // Also truncate `newValue` when `composingMaxLengthEnforced` is true.
+    return newValue.composing.isValid && !composingMaxLengthEnforced
+      ? newValue
+      : truncate(
+          newValue,
+          maxLength,
+          keepComposing: oldValue.composing.isValid && composingMaxLengthEnforced,
```

Comment:
Should we keep the composing region if the text length doesn't exceed `maxLength`?

Reply:
Start like this: